src/Libraries/Microsoft.Extensions.DataIngestion/Writers/VectorStoreWriter.cs

```diff
@@ -0,0 +1,201 @@
+    private const string KeyName = "key";
+    private const string EmbeddingName = "embedding";
+    private const string ContentName = "content";
+    private const string ContextName = "context";
+    private const string DocumentIdName = "documentid";
+
+    private readonly VectorStore _vectorStore;
+    private readonly int _dimensionCount;
+    private readonly VectorStoreWriterOptions _options;
+    private readonly bool _keysAreStrings;
+
+    private VectorStoreCollection<object, Dictionary<string, object?>>? _vectorStoreCollection;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="VectorStoreWriter{T}"/> class.
+    /// </summary>
+    /// <param name="vectorStore">The <see cref="VectorStore"/> to use to store the <see cref="IngestionChunk{T}"/> instances.</param>
+    /// <param name="dimensionCount">The number of dimensions that the vector has. This value is required when creating collections.</param>
+    /// <param name="options">The options for the vector store writer.</param>
+    /// <exception cref="ArgumentNullException">When <paramref name="vectorStore"/> is null.</exception>
+    /// <exception cref="ArgumentOutOfRangeException">When <paramref name="dimensionCount"/> is less or equal zero.</exception>
+    public VectorStoreWriter(VectorStore vectorStore, int dimensionCount, VectorStoreWriterOptions? options = default)
+    {
+        _vectorStore = Throw.IfNull(vectorStore);
+        _dimensionCount = Throw.IfLessThanOrEqual(dimensionCount, 0);
+        _options = options ?? new VectorStoreWriterOptions();
+
+        // Not all vector store support string as the key type, examples:
+        // Qdrant: https://github.com/microsoft/semantic-kernel/blob/28ea2f4df872e8fd03ef0792ebc9e1989b4be0ee/dotnet/src/VectorData/Qdrant/QdrantCollection.cs#L104
+        // When https://github.com/microsoft/semantic-kernel/issues/13141 gets released,
+        // we need to remove this workaround.
+        _keysAreStrings = vectorStore.GetType().Name != "QdrantVectorStore";
+    }
+
+    /// <summary>
+    /// Gets the underlying <see cref="VectorStoreCollection{TKey,TRecord}"/> used to store the chunks.
+    /// </summary>
+    /// <remarks>
+    /// The collection is initialized when <see cref="WriteAsync(IAsyncEnumerable{IngestionChunk{T}}, CancellationToken)"/> is called for the first time.
+    /// </remarks>
+    /// <exception cref="InvalidOperationException">The collection has not been initialized yet.
+    /// Call <see cref="WriteAsync(IAsyncEnumerable{IngestionChunk{T}}, CancellationToken)"/> first.</exception>
+    public VectorStoreCollection<object, Dictionary<string, object?>> VectorStoreCollection
+        => _vectorStoreCollection ?? throw new InvalidOperationException("The collection has not been initialized yet. Call WriteAsync first.");
+
+    /// <inheritdoc/>
+    public override async Task WriteAsync(IAsyncEnumerable<IngestionChunk<T>> chunks, CancellationToken cancellationToken = default)
```

Comment:
Are users typically going to read chunks asynchronously here?

Am noting that MEVD itself has batching Upsert, but which accepts IEnumerable rather than IAsyncEnumerable. We ideally want a single call to WriteAsync here to do a single call to Upsert on the MEVD provider, which will batch the upserts (single database roundtrip).